### PR TITLE
Implement BTB using the cache library

### PR DESCRIPTION
**Pull request by pranith**
This enables the BTB to be associative and use various replacement policies.

### CODE REVIEW

**Comment by pranith** [comment]
@giactra I tried using the templatized version of the cache library as an exercise. Can you please review this change? Thank you!

File: src/cpu/pred/btb_entry.hh
Comment (giactra) [inline src/cpu/pred/btb_entry.hh]: I wouldn't inherit from CacheEntry, as this is only adding complexity IMHO with no gain. It forces you to overload and hide some methods. The BTB entry should inherit from ReplaceableEntry

File: src/cpu/pred/btb_entry.hh
Comment (giactra) [inline src/cpu/pred/btb_entry.hh]: Nit: prefer using for consistency for what you are doing above with KeyType

File: src/cpu/pred/btb_entry.hh
Comment (giactra) [inline src/cpu/pred/btb_entry.hh]: Similarly to the BTBEntry discussion above. We need to pay the price and not inherit from SetAssociative. SetAssociative and the parent BaseIndexingPolicy are assuming an Addr key, whereas here our key is the Addr + Tid pair. I suggest you to have a look at what I did with TaggedIndexingPolicy... This is the hard bit I recon, as it requires you to define a templated SimObject. Happy to provide further assistance if needed

File: src/cpu/pred/btb_entry.hh
Comment (giactra) [inline src/cpu/pred/btb_entry.hh]: Ditto this is a workaround

File: src/cpu/pred/simple_btb.cc
Comment (giactra) [inline src/cpu/pred/simple_btb.cc:116]: With your change we are not checking if the btb entry is valid anymore. You need to check for validity in the match method

**Review by giactra — CHANGES_REQUESTED** [review]
Thanks @pranith for this change! As this is a purely refactoring change, could you run some workload with O3 and see if there are any stats difference between pre and after?

File: src/cpu/pred/simple_btb.cc
Comment (pranith) [reply src/cpu/pred/simple_btb.cc:116]: The CacheEntry's 'match' checks for the valid flag.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: I tried inheriting it from ReplaceableEntry and that just seems to be doing most of what CacheEntry is already doing for valid and other methods. We are duplicating most of the code here again.

Is there any reason why you prefer inheriting from ReplaceableEntry instead of CacheEntry and avoid duplication?

**Comment by pranith** [comment]
@giactra Thanks for the review! I tried inheriting from the base classes (instead of CacheEntry and IndexingPolicy) but it seems like we need to copy most of what these classes provide. I feel it becomes needlessly more labor intensive and error prone to copy the code and duplicate it in multiple places. 

Can you please provide the rationale for using the base classes? Only reason I can think of is that we are overriding base class methods with methods of a different signature, but I don't see what is wrong in doing this.

File: src/cpu/pred/btb_entry.hh
Comment (giactra) [reply src/cpu/pred/btb_entry.hh]: There are several reasons IMO

1) Design concern: Public virtual methods define the interface of a class. If you inherit from a class but you end up overloading such interface, then you are basically saying the BtbEntry is not really a CacheEntry, it is just reusing some of its code.

2) Possible bugs: this hiding can lead to potential bugs, see https://github.com/gem5/gem5/pull/1374

3) Tag ambiguity: the idea of defining a tag as the single address is just an artifact of past design. In this case the tag type is really the pair of addr and tid. But by using the CacheEntry you are basically defining two different tag types.

4) Performance: most of the BTBEntry logic is calling virtual functions. So you are needlessly adding entries to a VTABLE for a polymorphism that is never used, and you are calling them paying the vtable indirection penalty.

I understand that some code will look the same between the CacheEntry and the BTBEntry but IMHO the overhead is minimal and we would benefit a lot in decoupling the two entries, simplifying the design.

File: src/cpu/pred/btb_entry.hh
Comment (giactra) [inline src/cpu/pred/btb_entry.hh]: Please mark this as override

File: src/cpu/pred/simple_btb.cc
Comment (giactra) [reply src/cpu/pred/simple_btb.cc:116]: Ah right

File: src/cpu/pred/simple_btb.hh
Comment (giactra) [inline src/cpu/pred/simple_btb.hh]: It looks like this has been removed

File: src/cpu/pred/simple_btb.hh
Comment (odanrc) [inline src/cpu/pred/simple_btb.hh]: nit: re-add `// namespace gem5`

File: src/cpu/pred/simple_btb.hh
Comment (odanrc) [inline src/cpu/pred/simple_btb.hh]: I think this line exceeds the character limit (Lines must be a maximum of 79 characters long).

File: src/cpu/pred/simple_btb.hh
Comment (odanrc) [inline src/cpu/pred/simple_btb.hh]: nit: ´an a´

File: src/cpu/pred/simple_btb.hh
Comment (odanrc) [inline src/cpu/pred/simple_btb.hh]: <memory> is not used in this file

File: src/cpu/pred/simple_btb.cc
Comment (odanrc) [inline src/cpu/pred/simple_btb.cc]: Ditto line length

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh:160]: I think this function deserves an improvement in the documentation. Something to explain its difference from the insert() function. It seems to be a workaround due to the fact that the insertion only uses the key and the target and inst are not part of the key.

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh]: Since it seems that we only update() right after calling insert, the _tid assignment is redundant and should probably be dropped.

File: configs/common/cores/arm/HPI.py
Comment (odanrc) [inline configs/common/cores/arm/HPI.py]: Curiosity: where do the config numbers come from? (I am not too familiar with BTBs)

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: @giactra I tried your suggestion and created a BTBIndexingPolicy/BTBSetAssociative policy derived from SimObject. However, I seem to be running into a linking error. Can you please help?

File: configs/common/cores/arm/HPI.py
Comment (pranith) [reply configs/common/cores/arm/HPI.py]: It is approximately 'numEntires * entry_size' so that indexing is as before. I will re-work this part once before merging the final changes.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Ah, looks like I missed implementing the regenerateAddr method. It works now.

File: src/cpu/pred/simple_btb.hh
Comment (pranith) [reply src/cpu/pred/simple_btb.hh]: Done.

File: src/cpu/pred/simple_btb.hh
Comment (pranith) [reply src/cpu/pred/simple_btb.hh]: Fixed.

File: src/cpu/pred/simple_btb.hh
Comment (pranith) [reply src/cpu/pred/simple_btb.hh]: Fixed.

File: src/cpu/pred/simple_btb.hh
Comment (pranith) [reply src/cpu/pred/simple_btb.hh]: Removed.

File: src/cpu/pred/simple_btb.cc
Comment (pranith) [reply src/cpu/pred/simple_btb.cc]: Fixed.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Thanks for the explanation @giactra. BtbEntry is derived from CacheEntry means it has some properties of CacheEntry, but not all. This is I think a textbook example of OOP (mammals -> humans).

But, I am not really strong against implementing it from scratch. Just that, there will be code duplication.

I pushed a patch addressing all the review comments. Please take a look!

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Done.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Removed.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Done.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh:160]: Done.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Removed tid argument.

**Comment by pranith** [comment]
> Thanks @pranith for this change! As this is a purely refactoring change, could you run some workload with O3 and see if there are any stats difference between pre and after?

I reran with the patch and there were no stat differences in BTB. I updated the BTB params to keep the same behavior as before.

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [reply src/cpu/pred/btb_entry.hh:160]: I don´t see the change. Have you pushed it?

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh]: The tid should also be printed.

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh:190]: The rule of 3 dictates that if we need to define copy constructor then we should also need to define the copy assignment operator.

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh]: I think that it should instead return KeyType, since the tid *is* part of the tag. We need to make sure that what happened with the prefetchers (some use the secure bit, and others don´t) doesn´t happen in this class. We can already see an instance of the bug happening in this class´ print() and invalidate(), where the tid was forgotten.

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh]: Instead of storing it as an address, we should store it as a KeyType. This class should not use Addr.

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh]: As per the other comments, this should be the same visibility as the ´tag´ variable.

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh]: It should have used KeyType, because otherwise we end up forgetting about tid (see invalidate() using this function and not assigning a ´max tid´).

**Review by odanrc — CHANGES_REQUESTED** [review]

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh:160]: Sorry! missed pushing it earlier... pushed it in the latest commit.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh:190]: Added a copy assignment operator.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Changed the tag type to be KeyType througout.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Removed the independent tid member.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Updated.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Done.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Done.

**Review by odanrc — APPROVED** [review]

**Comment by pranith** [comment]
@odanrc Thaks for the review. I squashed all the commits into one. Not sure if this is can be done during merge?

**Comment by powerjg** [comment]
> I reran with the patch and there were no stat differences in BTB. I updated the BTB params to keep the same behavior as before.

I just want to double check that even though you saw there were no stats differences, did you make sure the BTB is being used in the tests? Our test coverage of many models like the BTB is pretty bad. You would likely have had to write your own test for this.

**Comment by pranith** [comment]
> 
> I just want to double check that even though you saw there were no stats differences, did you make sure the BTB is being used in the tests? Our test coverage of many models like the BTB is pretty bad. You would likely have had to write your own test for this.

The BTB sees the same number of accesses, hits, and, misses with the change for a small test program. So yes, I can confirm that it is being exercised and behaves as before the patch.

**Comment by odanrc** [comment]
> @odanrc Thaks for the review. I squashed all the commits into one. Not sure if this is can be done during merge?

I think for this small PR there will be no issue, but since there are other reviewers still reviewing I think it would have been better to wait for all approvals to be in order, in case someone was using the commits as part of their review methodology. IIRC, the merge does not trigger a merge squash, but I may be wrong.

**Comment by pranith** [comment]
@powerjg Can you please point me to docs on how to run the CI tests locally? It seems some quick boot tests are failing.

**Comment by powerjg** [comment]
See https://github.com/gem5/gem5/blob/stable/TESTING.md#running-tests-in-batch

**Comment by pranith** [comment]
@giactra Can you please review? This now passes CI tests and is ready for merge. Please let me know if you want me to squash the two commits together or you can do it during merge. Thanks!

File: src/cpu/pred/btb_entry.hh
Comment (giactra) [inline src/cpu/pred/btb_entry.hh]: Are you planning on extending this class? Otherwise I would consider devirtualizing these methods

File: src/cpu/pred/btb_entry.hh
Comment (giactra) [inline src/cpu/pred/btb_entry.hh]: Ditto

File: src/cpu/pred/btb_entry.hh
Comment (giactra) [inline src/cpu/pred/btb_entry.hh]: This looks expensive. can't we just assert (meaning we can skip this in the fast build) and simply check if set_idx is bigger or equal than sets.size()

This is what we were doing in the old version:

`    assert(btb_idx < numEntries);`

File: configs/common/cores/arm/HPI.py
Comment (giactra) [inline configs/common/cores/arm/HPI.py]: I feel your pain here. I understand you are carefully choosing size and entry_size to account for numEntries and the instShift.

It is 100% not your fault but I believe this is a bit fragile. What happens if someone changes numEntries in the BTB?

To overcome this problem I just created:

https://github.com/gem5/gem5/pull/1557

It will allow you to do

    btbIndexingPolicy = BTBSetAssociative(
        num_entries=numEntries, set_shift=instShiftAmt, assoc=associativity)

Would you be up for waiting for my PR to be merged?

**Review by giactra — COMMENTED** [review]
Hi @pranith I have some minor comments and a more important one that relates to my new PR: https://github.com/gem5/gem5/pull/1557

(I developed the PR once I saw this one here and its problems)

File: configs/common/cores/arm/HPI.py
Comment (pranith) [reply configs/common/cores/arm/HPI.py]: Thanks for the PR @giactra. It definitely makes it much easier to instantiate the structures. I will rebase on top of your pull request and update this patch.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Updated.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Removed virtual.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Updated as suggested.

File: configs/common/cores/arm/HPI.py
Comment (pranith) [reply configs/common/cores/arm/HPI.py]: Updated accordingly.

**Comment by pranith** [comment]
@andysan Now that we have the tag based addressing in the cache, can you please review the 'tagMask' commit of this PR? We need tagMask for the BTB (which I just realized).Thanks!

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh]: Nit: this is a tag type - i.e., not plural

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh]: Fix to fit in line length

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh]: Is this function being called and its output verified anywhere? At a glance this does not seem to be correct because it lacks the xor on the tid

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [inline src/cpu/pred/btb_entry.hh]: This should probably be a const function - I don´t think we ever expect to change the class when checking for a match

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Updated.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Fixed.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: I don't think we use this function. I replaced the body with an 'assert(false)'. It would be good to remove it from the base class and only define it for mem-cache related policies, but it didn't seem straight forward.

File: src/cpu/pred/btb_entry.hh
Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: Updated.

File: src/cpu/pred/btb_entry.hh
Comment (odanrc) [reply src/cpu/pred/btb_entry.hh]: Regarding not having it, if you are sure it is not going to be needed, then sure we don´t need to implement it. One alternative is to make the base class no longer pure virtual and panic there. Speaking of which, it would be better to use something in the lines of ´panic(not implemented)´ than the assertion, as it is not a user error, but a lack of feature.

Comment (pranith) [reply src/cpu/pred/btb_entry.hh]: I don't know what to provide in the default non-pure virtual function, so I just updated the assert() to a panic.

**Comment by andysan** [comment]
> @andysan Now that we have the tag based addressing in the cache, can you please review the 'tagMask' commit of this PR? We need tagMask for the BTB (which I just realized).Thanks!

My main concern with the tag bits param is that there is no way for users to reason about this. Does setting it to something other than 64 truncate the TID or the address? What if we add even more information (we'll need to do this if we want to be representative of real hardware) in the tags does it truncate the same fields as before? I think realistically, we can't expect users to be able to use this param.

A better solution would be to deal with truncation in the btb indexing policy instead where the semantic information is available. That'd let users decide how many bits of address to include, whether to include the TID, VMID, etc., which would be much more meaningful than just truncating some bits.

I'm currently out of office, so I won't have time to look at this in more detail. @giactra could you have a look and drive this from our side?

**Comment by pranith** [comment]
> A better solution would be to deal with truncation in the btb indexing policy instead where the semantic information is available. That'd let users decide how many bits of address to include, whether to include the TID, VMID, etc., which would be much more meaningful than just truncating some bits.

I updated BTB indexing policy to use tag mask independent of the base policy. There are other users of cache library that need similar tag mask like bop prefetcher, but I left it alone for now.

> 
> I'm currently out of office, so I won't have time to look at this in more detail. @giactra could you have a look and drive this from our side?

@giactra Can you please review? Thanks!

**Comment by pranith** [comment]
@giactra @odanrc Can you please review this updated patch? It would be good to have it merged if there are no new comments.

File: src/cpu/pred/BranchPredictor.py
Comment (odanrc) [inline src/cpu/pred/BranchPredictor.py:93]: How does this line work? Is this string a description of the template param?

**Comment by odanrc** [comment]
@pranith I cannot spot the differences between the previous version and the new version. This is probably because the commits were squashed together. Could you point to where we should focus the review efforts? As a side note, in the future I recommend avoiding squashing new commits added to a change that was already under review, as it makes spotting the changes much harder.

**Comment by pranith** [comment]
> @pranith I cannot spot the differences between the previous version and the new version. This is probably because the commits were squashed together. Could you point to where we should focus the review efforts? As a side note, in the future I recommend avoiding squashing new commits added to a change that was already under review, as it makes spotting the changes much harder.

I apologize for this. There were some conflicts in the merge and I had to rebase to fix the conflicts. I will add commits addressing review commits only in the future.

The main change from your last review is the addition of the 'tagMask' field in the BTBIndexingPolicy. Rest of it remains the same.

File: src/cpu/pred/BranchPredictor.py
Comment (pranith) [reply src/cpu/pred/BranchPredictor.py:93]: I copy+pasted this from the 'BaseIndexingPolicy' from 'mem/cache/tags/indexing_policies/IndexingPolicies.py' since I am trying to reuse the IndexingPolicyTemplate just like it does. Please let me know if this needs to be updated.

File: src/cpu/pred/BranchPredictor.py
Comment (giactra) [inline src/cpu/pred/BranchPredictor.py]: Is this a leftover from the initial version?

**Review by giactra — CHANGES_REQUESTED** [review]
Hi @pranith apologies for the late review. I only have one comment left. Then if @odanrc is also happy I will approve the PR

File: src/cpu/pred/BranchPredictor.py
Comment (pranith) [reply src/cpu/pred/BranchPredictor.py]: Indeed, removed the line. I will squash the commits if you think this is OK. Thanks!

Comment (giactra) [reply src/cpu/pred/BranchPredictor.py]: Yes feel free to squash

Comment (pranith) [reply src/cpu/pred/BranchPredictor.py]: Done.

**Review by odanrc — APPROVED** [review]

**Review by giactra — APPROVED** [review]

**Pull request merged**
